Integer comparison(#23518)

u7122029 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #23518 


#### Brief description of what is fixed or changed
Modified relational simplifier such that if both sides are numeric and nonzero, we attempt to simplify `Rel(RHS - LHS, 0, op)`, then attempt to simplify the original relational expression. The simplified expression returned will be the "better" result of the two i.e: the respective simplified expression with lower measure.

Also added tests for the modification.

#### Other comments
None.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* simplify
  * 

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* simplify
  * Added a strategy where the simplifier will consider subtracting both sides of a numeric, relational expression.
<!-- END RELEASE NOTES -->

u7122029 committed:
Update .mailmap

(cherry picked from commit 49a3741c4be26e9235fb96c683631858ecea2f2b)

u7122029 committed:
Written solution for (#23518)

u7122029 committed:
Corrected tests for new method (#23518)

u7122029 committed:
Slight change to choosing best simplification. (#23518)

u7122029 committed:
Merge branch 'master' of https://github.com/sympy/sympy into IntegerComparison(#23518)

 Conflicts:
	doc/src/guides/solving/index.md

u7122029 committed:
Removed trailing newline.

u7122029 committed:
Removed unused imports.

u7122029 committed:
Amended new simplification method to only handle NumberKind expressions.

u7122029 committed:
Fixed type check.

u7122029 committed:
Update .mailmap

u7122029 committed:
Remove function call in test file.

oscarbenjamin commented on .mailmap:
```diff
@@ -1500,6 +1500,8 @@ tborisova <ts.borisova3@gmail.com> <ts_borisova@abv.bg>
 tools4origins <tools4origins@gmail.com>
 tttc3 <T.Coxon2@lboro.ac.uk>
+u7122029 <111028268+u7122029@users.noreply.github.com>
+u7122029 <u7122029@anu.edu.au>
```
There should not be separate entries for the same person:
https://github.com/sympy/sympy/blob/5e50b55cc3ce0731574397d00628e429e23b5e53/.mailmap#L138-L148

u7122029 replied:
Just fixed. Sorry about that, I'm sure you've had to say the same thing to so many other contributors.

u7122029 committed:
Fixed unrelated authors.

asmeurer commented on doc/src/guides/solving/solve-system-of-equations-algebraically.md:
```diff
@@ -1,29 +1,15 @@
-\sqrt{6z}, y = -4z)\}}$.
+Use SymPy to solve a system of equations algebraically. For example, solving
+$x^2 + y = 2z, y = -4z$ for x and y (assuming z is a constant or parameter)
+yields $\{(x = -\sqrt{6z}, y = -4z),$ ${(x = \sqrt{6z}, y = -4z)\}}$.
```
Why is this file modified? Did you resolve a merge conflict incorrectly?

u7122029 replied:
It's possible, but I'm not sure. My repo before the forced merge was about 192 commits behind the main repo.

u7122029 replied:
Just copied the contents of the file in master and pasted it into the same file in my branch. Does this fix this issue?

u7122029 committed:
Merge branch 'master' of https://github.com/sympy/sympy into IntegerComparison(#23518)

u7122029 committed:
Corrected possibly incorrect merge conflict.

u7122029 committed:
Merge remote-tracking branch 'origin/IntegerComparison(#23518)' into IntegerComparison(#23518)

oscarbenjamin commented:
I'm not sure why the authors test is failing:
```
This author is not included in the .mailmap file:
Sam Brockie <sambrockie@icloud.com>
```
That seems unrelated here...

u7122029 commented:
> I'm not sure why the authors test is failing:
> 
> ```
> This author is not included in the .mailmap file:
> Sam Brockie <sambrockie@icloud.com>
> ```
> 
> That seems unrelated here...

I'm not sure either...

u7122029 committed:
Added Sam Brockie to .mailmap to resolve authors PR CI fail.

u7122029 commented:
Just made a commit to the repo to correct the missing author and it seems the authors test has passed as of now.

brocksam commented:
> I'm not sure why the authors test is failing:
> 
> ```
> This author is not included in the .mailmap file:
> Sam Brockie <sambrockie@icloud.com>
> ```
> 
> That seems unrelated here...

This was almost certainly my fault. It has been addressed in PR #24209, which is now merged.

> Just made a commit to the repo to correct the missing author and it seems the authors test has passed as of now.

You can probably revert https://github.com/sympy/sympy/pull/24198/commits/74742a281857d16168def099467749880df6acca and we can trigger the authors check again.

u7122029 committed:
Revert "Added Sam Brockie to .mailmap to resolve authors PR CI fail."

This reverts commit 74742a281857d16168def099467749880df6acca.

u7122029 commented:
> > I'm not sure why the authors test is failing:
> > ```
> > This author is not included in the .mailmap file:
> > Sam Brockie <sambrockie@icloud.com>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That seems unrelated here...
> 
> This was almost certainly my fault. It has been addressed in PR #24209, which is now merged.
> 
> > Just made a commit to the repo to correct the missing author and it seems the authors test has passed as of now.
> 
> You can probably revert [74742a2](https://github.com/sympy/sympy/commit/74742a281857d16168def099467749880df6acca) and we can trigger the authors check again.

Just reverted and pushed. Let's see what happens.

u7122029 commented:
Tests have passed.

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -23,6 +23,8 @@
 from .containers import Tuple
 from .symbol import Symbol
+from .add import Add
+from sympy.core.kind import NumberKind
```
Does it not work to do `from .kind import` here?

u7122029 replied:
I was previously using VSCodium which seems to be a slower IDE compared to PyCharm at checking package dependencies. So I had written the import like that. I've changed the import to `from .kind import NumberKind` as you suggested.

oscarbenjamin commented on sympy/simplify/tests/test_simplify.py:
```diff
@@ -1046,3 +1049,21 @@ def test_issue_22210():
+
+    expr = M*N >= N
+    expr = expr.simplify()
+    assert expr == True
```
Of the added tests this is the only one that doesn't already pass on master. I'm struggling to understand why there are so many code changes when for this example it could be as simple as this:
```python
In [19]: simplify(expr.rhs - expr.lhs).is_nonnegative
Out[19]: True
```

u7122029 replied:
As far as I can see, this test does pass on my local branch. Have there been other changes to the simplification function?

u7122029 replied:
I believe it is also more convenient to simplify an expression without needing to check other properties about it manually, so I thought there was value in solving this issue.

oscarbenjamin replied:
My point is this: you have changed a lot of code and I'm finding it difficult when I look in the diff to understand precisely why those changes have been made so I looked at the tests. Most of the tests already pass on master without the changes in this PR. This is the only one that this PR actually improves but then I don't understand why the changes in this PR are as complicated as they are in order to accommodate what seems like a fairly simple case.

u7122029 replied:
I had introduced `_eval_simplification()` as a helper function because the code inside it used to be in `_eval_simplify()` (both functions in `relational.py`) and it looked very bloated and difficult to understand. I therefore wrote `_eval_simplification()` to split the functionality of `_eval_simplify()` into separate, smaller parts so my solution would be easier to implement and understand.

u7122029 replied:
It does not work because of line 469 `m = linear_coeffs(dif, *free)` (I'm not sure how to reference code directly). The problem here is that it tries to get linear coefficients of all free variables (M and N in our example), however, `M*N - N` (the difference) has co-dependent coefficients. So it stops computation and returns the expression as-is. Do you believe I could move my code into `_eval_simplification` as another try-except block underneath this one?

oscarbenjamin replied:
We should avoid using `try/except` but I would have thought this logic should somehow be handled in the block that treats `dif` as a polynomial.

u7122029 replied:
Where is this block?

oscarbenjamin replied:
Oh, I see that block only happens if there is a single free symbol which isn't the case here.

Let me put this another way: the code that you are editing is already quite convoluted and complicated and your changes here seem to add more complication on top. Actually I suspect that the best approach is to somehow simplify the logic in the existing code rather than just adding more on top.

u7122029 replied:
I don't think I added any complications on top. I literally allowed the old algorithm to run (which I didn't change), then my new simple algorithm to run. I return the expression that is better out of the two.

oscarbenjamin replied:
There are many different kinds of complexity that can be considered:

- Computational complexity - having two algorithms potentially doubles the computational cost.
- Code complexity - having multiple branches makes reasoning about the code more complicated especially when try/except is used because that adds many implicit branches.
- Functional complexity - if there are different branches that apply in slightly different cases then how does that interact with the different kinds of inputs? One branch is for Expr only and the other is for NumberKind only: why do they have different conditions?
- etc

It is almost always better to try to improve the existing code rather than to just add more code on top. The thing to do is study the existing code and consider:

- What is it trying to do?
- In what cases does it work as intended?
- Why does it not work as intended in the case that you are interested in?
- Is there a simpler way of handling all cases of interest?

When I look at the code in the `len(free) >= 2` block I can see that it only attempts to handle linear equations. But I can also see that while it's quite complicated it is mostly just doing something quite simple:

- It factors out `scale` which could just be done more simply using `Mul.make_args(factor_terms(dif))`.
- It then expects to have an `Add` leftover and it will take some terms of the `Add` and move them to the other side. This just needs to use `Add.make_args` and check which terms don't involve the free symbols.

There is no reason that this should only work in the linear case and could not also apply to polynomials and I'm fairly sure that the code could be changed to do this more simply and handle both linear and polynomial inequalities.

u7122029 replied:
Thank you so much for such a detailed explanation of the situation and how I can make my way around it. I just pushed to my fork. The sun is about to rise where I am so I will check the CI next morning to see if everything passed.

u7122029 committed:
Implemented suggestion from @oscarbenjamin.

u7122029 committed:
Fixed unused NumberKind import.

u7122029 commented:
The good news is that most of the unit tests passed.

The bad news is that the ones that failed did so because of differences in representation as far as I could see.

oscarbenjamin commented:
> The bad news is that the ones that failed did so because of differences in representation as far as I could see.

That's not necessarily a problem. The output of simplify is only heuristically defined so there isn't always a clear way to decide what is "correct".

oscarbenjamin commented on sympy/core/relational.py:
```diff
@@ -412,92 +415,95 @@ def equals(self, other, failing_expression=False):
+                    r = r.func(Poly.from_list(c, x).as_expr(), -constant / scale)
+                except PolynomialError:
+                    pass
+        elif len(free) >= 2:
+            from sympy import factor_terms
+            dif = r.lhs - r.rhs
+
+            factored = Mul.make_args(factor_terms(dif))
```
I think it's worth considering whether there is actually any need to have different branches depending on `len(free)` at all. Do the different branches actually do fundamentally different things?

oscarbenjamin commented:
I think it would be better not to have the `_eval_simplification` method. I agree that `_eval_simplify` is unnecessarily complicated but moving it to a separate function doesn't really help that. It would be better just to simplify the actual code.

u7122029 commented:
Good morning Oscar, thanks for following up. I wanted to make the initial push to see if my code still works on all the test cases. Do you know how to run specific test cases/test files through the terminal so I don't have to keep pushing? 

Should I also change the test cases that failed so they line up with the output of my new solution?

u7122029 committed:
Removed _eval_simplification

u7122029 committed:
Added more robust changes.

u7122029 committed:
Added collateral.py to show differences in test cases. Also modified relational.py to check if difference is zero.

u7122029 commented:
I added a file called `collateral.py` so you can see the test cases that failed and their slight differences. I also modified the simplifier so that it will check if the difference is zero since without it, the simplifier would not evaluate complex expressions with no free vars.

oscarbenjamin commented:
> I added a file called `collateral.py`

Please don't add new files even temporarily. The commits will need to be squashed so that the file does not exist in the git history.

u7122029 commented:
> > I added a file called `collateral.py`
> 
> Please don't add new files even temporarily. The commits will need to be squashed so that the file does not exist in the git history.

Got it. That aside, should I change the tests that are failing so that they will work for the new simplifier?

oscarbenjamin commented:
> ```python
> File "/home/runner/work/sympy/sympy/sympy/core/tests/test_relational.py", line 727, in test_simplify_relational
> Before: -2*q + 2*r <= 0
> Expected: q >= r
> Actual: q >= r, count_ops: 1
> ```

It looks like the expected and actual are the same here. When I check with master I get the same result `q >= r` as the PR.

> ```
> File "/home/runner/work/sympy/sympy/sympy/core/tests/test_relational.py", line 1143, in test_polynomial_relation_simplification
> Before: 3*x*(x + 1) + 4 >= 3*x
> Expected: [x**2 >= -4/3, -x**2 <= 4/3], count_ops: [4, 4]
> Actual: 3*x**2 >= -4, count_ops: 4
> ```

Here it would be better to divide through by 3.

> ```
> File "/home/runner/work/sympy/sympy/sympy/core/tests/test_relational.py", line 1156, in test_multivariate_linear_function_simplification
> Before: a + b + c + 2*d < 3*d - f + g
> Expected: a < -b - c + d - f + g, count_ops: 5
> Actual: d + g > a + b + c + f, count_ops: 5
> ```

I think that the PR s an improvement for this case.

> ```
> File "/home/runner/work/sympy/sympy/sympy/functions/elementary/tests/test_piecewise.py", line 454, in test_issue_22917
> Before: Piecewise((2, ((x > y + 1) | (2*x - 2*y > 1)) & ((x > y + 1) | (x - y > 1) | (2*x - 2*y > 1))), (2*y + 2, (x > y + 1) | (2*x - 2*y > 1)), (4, (x - y > 1) & ((x > y + 1) | (x - y > 1) | (2*x - 2*y > 1))), (2*y + 4, x - y > 1), (4*x, (x > y + 1) | (x - y > 1) | (2*x - 2*y > 1)), (4*x + 2*y, True))
> Expected: Piecewise((2, (x - y > 1/2) | (x - y > 1)), (2*y + 4, x - y > 1), (4*x + 2*y, True)), count_ops: 11
> Actual: Piecewise((2, (x > y + 1) | (2*x - 2*y > 1)), (4, x > y + 1), (4*x + 2*y, True)), count_ops: 10
> ```

I'm not sure I understand what's happening here:
```
In [4]: Piecewise((2, (x - y > 1/2) | (x - y > 1)), (2*y + 4, x - y > 1), (4*x + 2*y, True))
Out[4]: 
⎧    2      for x - y > 0.5 ∨ x - y > 1
⎪                                      
⎨ 2⋅y + 4          for x - y > 1       
⎪                                      
⎩4⋅x + 2⋅y           otherwise         

In [5]: Piecewise((2, (x > y + 1) | (2*x - 2*y > 1)), (4, x > y + 1), (4*x + 2*y, True))
Out[5]: 
⎧    2      for x > y + 1 ∨ 2⋅x - 2⋅y > 1
⎪                                        
⎨    4              for x > y + 1        
⎪                                        
⎩4⋅x + 2⋅y            otherwise  
```
Why has the `2*y + 4` become `4`?

> ```
> File "/home/runner/work/sympy/sympy/sympy/integrals/tests/test_integrals.py", line 1121, in test_issue_4527
> Before: Piecewise((0, Eq(k, 0) | Eq(m, 0) | (Eq(k, 0) & Eq(m, 0)) | (Eq(k, 0) & Eq(k, m) & Eq(m, 0)) | (Eq(k, 0) & Eq(m, 0) & Eq(k, -m)) | (Eq(k, 0) & Eq(k, m) & Eq(m, 0) & Eq(k, -m))), (-pi/2, Eq(k, -m) | (Eq(k, 0) & Eq(k, -m)) | (Eq(k, m) & Eq(k, -m)) | (Eq(m, 0) & Eq(k, -m)) | (Eq(k, 0) & Eq(k, m) & Eq(k, -m)) | (Eq(k, m) & Eq(m, 0) & Eq(k, -m))), (pi/2, Eq(k, m) | (Eq(k, 0) & Eq(k, m)) | (Eq(k, m) & Eq(m, 0))), (0, True))
> Expected: Piecewise((0, Eq(k, 0) | Eq(m, 0)), (-pi/2, Eq(k, -m) | (Eq(k, 0) & Eq(m, 0))), (pi/2, Eq(k, m) | (Eq(k, 0) & Eq(m, 0))), (0, True)), count_ops: 10
> Actual: Piecewise((0, Eq(k, 0) | Eq(m, 0)), (-pi/2, Eq(k + m, 0) | (Eq(k, 0) & Eq(m, 0))), (pi/2, Eq(k, m) | (Eq(k, 0) & Eq(m, 0))), (0, True)), count_ops: 10
> ```
I think this one is fine.

> ```
> File "/home/runner/work/sympy/sympy/sympy/logic/tests/test_boolalg.py", line 316, in test_simplification_boolalg
> Before: (x <= y) | (z & (y > x))
> Expected: x <= y, count_ops: 1
> Actual: x <= y, count_ops: 1
> ```

What is the difference here?

> ```
> File "/home/runner/work/sympy/sympy/sympy/logic/tests/test_boolalg.py", line 1111, in test_relational_simplification
> Before: Eq(x, y) & Eq(x, -y)
> Expected: Eq(x, 0) & Eq(y, 0), count_ops: 3
> Actual: Eq(x, y) & Eq(x + y, 0), count_ops: 4
> ```

The previous result was better.

> ```
> Before: -2*(1 + 2**(pi/2) + 3**pi)**(1/pi) + (2**pi + 2**(3*pi/2) + 6**pi)**(1/pi) <= 0
> Expected: True, count_ops: 0
> Actual: True, count_ops: 0
> ```

I don't understand what the difference is here.

oscarbenjamin commented:
> That aside, should I change the tests that are failing so that they will work for the new simplifier?

Changing the tests until they pass is a better way of demonstrating to anyone reviewing this what the effect of the change is even if we would later change the code and the tests again.

u7122029 commented:
> > ```python
> > File "/home/runner/work/sympy/sympy/sympy/core/tests/test_relational.py", line 727, in test_simplify_relational
> > Before: -2*q + 2*r <= 0
> > Expected: q >= r
> > Actual: q >= r, count_ops: 1
> > ```
> 
> It looks like the expected and actual are the same here. When I check with master I get the same result `q >= r` as the PR.

I had this here initially because the code used to output `q - r >= 0`. I fixed it. I'm going to use the same reasoning for all the other test cases that ended up with the actual being the same as the expected.

> > ```
> > File "/home/runner/work/sympy/sympy/sympy/core/tests/test_relational.py", line 1143, in test_polynomial_relation_simplification
> > Before: 3*x*(x + 1) + 4 >= 3*x
> > Expected: [x**2 >= -4/3, -x**2 <= 4/3], count_ops: [4, 4]
> > Actual: 3*x**2 >= -4, count_ops: 4
> > ```
> 
> Here it would be better to divide through by 3.

I think for this I can check if there is one free variable term on one side, and divide its coefficient on both sides. But hang on, that's what the previous code did...
> > ```
> > File "/home/runner/work/sympy/sympy/sympy/core/tests/test_relational.py", line 1156, in test_multivariate_linear_function_simplification
> > Before: a + b + c + 2*d < 3*d - f + g
> > Expected: a < -b - c + d - f + g, count_ops: 5
> > Actual: d + g > a + b + c + f, count_ops: 5
> > ```
> 
> I think that the PR s an improvement for this case.

Good to hear. Please pretend I said the same thing for all other test cases which you think are fine.

> > ```
> > File "/home/runner/work/sympy/sympy/sympy/functions/elementary/tests/test_piecewise.py", line 454, in test_issue_22917
> > Before: Piecewise((2, ((x > y + 1) | (2*x - 2*y > 1)) & ((x > y + 1) | (x - y > 1) | (2*x - 2*y > 1))), (2*y + 2, (x > y + 1) | (2*x - 2*y > 1)), (4, (x - y > 1) & ((x > y + 1) | (x - y > 1) | (2*x - 2*y > 1))), (2*y + 4, x - y > 1), (4*x, (x > y + 1) | (x - y > 1) | (2*x - 2*y > 1)), (4*x + 2*y, True))
> > Expected: Piecewise((2, (x - y > 1/2) | (x - y > 1)), (2*y + 4, x - y > 1), (4*x + 2*y, True)), count_ops: 11
> > Actual: Piecewise((2, (x > y + 1) | (2*x - 2*y > 1)), (4, x > y + 1), (4*x + 2*y, True)), count_ops: 10
> > ```
> 
> I'm not sure I understand what's happening here:
>
> ```
> In [4]: Piecewise((2, (x - y > 1/2) | (x - y > 1)), (2*y + 4, x - y > 1), (4*x + 2*y, True))
> Out[4]: 
> ⎧    2      for x - y > 0.5 ∨ x - y > 1
> ⎪                                      
> ⎨ 2⋅y + 4          for x - y > 1       
> ⎪                                      
> ⎩4⋅x + 2⋅y           otherwise         
> 
> In [5]: Piecewise((2, (x > y + 1) | (2*x - 2*y > 1)), (4, x > y + 1), (4*x + 2*y, True))
> Out[5]: 
> ⎧    2      for x > y + 1 ∨ 2⋅x - 2⋅y > 1
> ⎪                                        
> ⎨    4              for x > y + 1        
> ⎪                                        
> ⎩4⋅x + 2⋅y            otherwise  
> ```
> 
> Why has the `2*y + 4` become `4`?

If you have a look at the original piecewise function, the case there `2y + 4` had a condition that pretty much simplified down to `x > y + 1`. This is the same condition you would end up with for the case where the output is `4`. It also occurs earlier in the list of cases, so technically the simplifier is correct. I have an image of the original, previous, and new simplifications below:

![latec](https://user-images.githubusercontent.com/111028268/200169470-fb5cce18-d100-4b56-b3b2-f4ccb2abbb47.png)

> > ```
> > File "/home/runner/work/sympy/sympy/sympy/logic/tests/test_boolalg.py", line 1111, in test_relational_simplification
> > Before: Eq(x, y) & Eq(x, -y)
> > Expected: Eq(x, 0) & Eq(y, 0), count_ops: 3
> > Actual: Eq(x, y) & Eq(x + y, 0), count_ops: 4
> > ```
>
> The previous result was better.

I'm going to try and see what the previous code did with this test case.

> > ```
> > Before: -2*(1 + 2**(pi/2) + 3**pi)**(1/pi) + (2**pi + 2**(3*pi/2) + 6**pi)**(1/pi) <= 0
> > Expected: True, count_ops: 0
> > Actual: True, count_ops: 0
> > ```
> 
> I don't understand what the difference is here.

This one is particularly interesting because the new simplifier was unable to reduce the left hand side to 0 until I introduced a bit of code from the old implementation back into the simplifier.

> > That aside, should I change the tests that are failing so that they will work for the new simplifier?
> 
> Changing the tests until they pass is a better way of demonstrating to anyone reviewing this what the effect of the change is even if we would later change the code and the tests again.

Now that I think of it. I can see why now - it's because you can directly see the changes through this GitHub website and more easily reference bits of code. Wow that's actually impressive.

oscarbenjamin commented:
> If you have a look at the original piecewise function, the case there `2y + 4` had a condition that pretty much simplified down to `x > y + 1`

Oh, I see: the middle case never applies because if its condition is true then the first case applies. Ideally the simplifier would just remove that case.

oscarbenjamin commented:
> you can directly see the changes through this GitHub website and more easily reference bits of code

Yes, and also if I can see that GitHub has verified that the tests have passed then I know that I don't necessarily need to try anything out on my computer. The changes to the tests have already been verified so I only need to inspect the changed code.

u7122029 commented:
> > If you have a look at the original piecewise function, the case there `2y + 4` had a condition that pretty much simplified down to `x > y + 1`
> 
> Oh, I see: the middle case never applies because if its condition is true then the first case applies. Ideally the simplifier would just remove that case.

Yeah that's true, however, could we maybe make a new issue for that so we can focus on that later? I feel like the original output and the current one are similar in complexity. I think I'd also have to deal with another simplifier for piecewise functions so that would be another can of worms that we would be opening.

oscarbenjamin commented:
> could we maybe make a new issue for that so we can focus on that later?

It's fine to leave that for later. I think @oscargus already has a PR somewhere that probably handles that case.

u7122029 committed:
Next attempt at simplifier

u7122029 committed:
Reset. Try solution again for expressions with one free var only.

u7122029 committed:
Corrected negative scale division.

u7122029 commented:
Thanks for guiding me along so far. Right now I've decided to reset back to the old code because I found that I was writing really complicated code and losing my vision of what we really wanted from this PR. I started by working with expressions with only one variable.

oscarbenjamin commented:
> I found that I was writing really complicated code and losing my vision of what we really wanted from this PR.

Okay but what I want if this code is to be changed is for the code to be simplified and certainly not made more complex. Currently this PR makes it more complicated for what seem like quite marginal improvements.

I'm pretty sure that all of this code (all branches) can be replaced with something much simpler along the lines of what I referred to above so I would like the number of branches reduced rather than increased.

oscarbenjamin commented:
By the way part of the difficulty I have with this PR is the fact that you have moved lots of code into a separate function. That makes the diff hard to interpret because I can't easily see what is being changed and what is just being moved. It is generally better to handle a refactoring separately (i.e. in a separate PR) from any change to behaviour.

Also while you have removed the file that was added it is not enough to add a commit that removes the file after a commit that adds the file. The file cannot exist in any of the commits so the commits need to be squashed.

oscarbenjamin commented:
I took a look at `Relational._eval_simplify` and made some notes:
```python
        # 1. simplify both sides and reconstruct
        # 2. if not relational any more then return
        # 3. if either side is not Expr return
        # 4. subtract both sides
        # 5. if numeric then evaluate diff numerically
        # 6. if diff.equals(0) then diff = 0
        #
        # 7. only one symbol:
        # 7a. if linear then make symbol the lhs if sign of coefficient is known.
        # 7b. if coeff is zero then remove that term.
        # 7c. if polynomial then separate constant and make primitive.
        #
        # 8. two or more symbols
        # 8a. make a linear polynomial in the symbols
        # 8b. move constant term to rhs if nonzero
        # 8b'. keep only "first" term on lhs if constant = 0
        # 8c. make primitive
        # 8d. put the first term on the lhs (why?)
        #
        # 9. Use measure to decide if simplified expression is better.
        #
        # Goals:
        # 1. Evaluate if possible
        # 2. Divide common factor of coefficients of "symbols".
        # 3. Put symbols on the lhs and constant on the rhs.
        # 4. Make canonical somehow
        #
        # Notes:
        # - 8b' is questionable.
        # - Using poly causes expansion of polynomial powers
        # - gcd always returns a positive number?
        # - Only checks for symbols that have is_real not False
        # - Fails for expressions like simplify(2*exp(x) + 2*y > 2)
        #
        # Questions:
        # - Should some terms be moved to rhs using could_extract_minus_sign?
        # (currently done only in the linear case).
        # - Is polynomial power expansion a good idea?
        # - Rule 2, 5 and 6 if r is not Relational then surely it is just True
        # or False.
```
I think that the first thing to decide is what in general the expected output is supposed to be. The rules here look a bit arbitrary. Also the fact that `measure` is used means that we can only heuristically expect any feature of the output so there is no guarantee that simplify will place a relational into a canonical form.

u7122029 commented:
Btw what exactly is commit squashing? This is the first time I've ever heard of it and I don't want to squash the wrong commits together, making a bigger mess.

u7122029 commented:
![difficulty](https://user-images.githubusercontent.com/111028268/200217052-b8db969b-4b69-42be-8da0-f68c63992741.png)
All the commits in green are the ones I chose to squash. However, this action has resulted in a commit on a separate branch. I tried rebasing and pushing but all the commits from online just come back. What do I do from here?

oscargus commented:
The purpose of squashing commits is to reduce the number of commits (currently 25) either to a single commit or to a few commits each with a well-defined purpose.

Not sure what you have done now, but the standard way is to do something like

`git rebase -i HEAD~25`

which will get you 25 commits back.

You will now get a list of the last 25 commits and can select squash for those that you want to squash (or probably easier: fixup, as you may not want to keep the commit messages). There are also a few other options, like rewording commit messages.

If you remove a commit from the list, it will be completely lost. If you remove all commits, you will abort the rebase.

Once you have rebased/squashed, you will need to push with `--force-with-lease`.

oscarbenjamin commented:
If you haven't tried squashing commits before then you might find it easier just to start again with a fresh PR from the master branch.

oscarbenjamin commented:
I'm going to mark this PR as draft for now. Feel free to mark it as ready for review once some changes have been made.
